File: ios/Classes/Presentation/Handler/MethodCallHandler.swift
import Flutter
import Foundation

/// Handler for method calls from Flutter
class MethodCallHandler {
    private let repository: TaskRepositoryImpl
    private let createRequestUseCase: CreateRequestUseCase
    private let getRequestStatusUseCase: GetRequestStatusUseCase
    private let getResponseUseCase: GetResponseUseCase
    private let cancelRequestUseCase: CancelRequestUseCase
    private let deleteRequestUseCase: DeleteRequestUseCase
    
    init() {
        self.repository = TaskRepositoryImpl()
        self.createRequestUseCase = CreateRequestUseCase(repository: repository)
        self.getRequestStatusUseCase = GetRequestStatusUseCase(repository: repository)
        self.getResponseUseCase = GetResponseUseCase(repository: repository)
        self.cancelRequestUseCase = CancelRequestUseCase(repository: repository)
        self.deleteRequestUseCase = DeleteRequestUseCase(repository: repository)
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "createRequest":
            handleCreateRequest(call: call, result: result)
        case "getRequestStatus":
            handleGetRequestStatus(call: call, result: result)
        case "getBatchRequestStatus":
            handleGetBatchRequestStatus(call: call, result: result)
        case "getResponse":
            handleGetResponse(call: call, result: result)
        case "cancelRequest":
            handleCancelRequest(call: call, result: result)
        case "deleteRequest":
            handleDeleteRequest(call: call, result: result)
        case "getPendingTasks":
            handleGetPendingTasks(call: call, result: result)
        case "cancelAllTasks":
            handleCancelAllTasks(call: call, result: result)
        // New methods for queue management
        case "getQueueStats":
            handleGetQueueStats(call: call, result: result)
        case "setMaxConcurrentTasks":
            handleSetMaxConcurrentTasks(call: call, result: result)
        case "setMaxQueueSize":
            handleSetMaxQueueSize(call: call, result: result)
        case "syncQueueState":
            handleSyncQueueState(call: call, result: result)
        case "processQueue":
            handleProcessQueue(call: call, result: result)
        case "getPendingCompletedTasks":
            handleGetPendingCompletedTasks(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func handleCreateRequest(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Request data is required", details: nil))
            return
        }
        
        Task {
            do {
                let request = RequestMapper.fromFlutterMap(args)
                let taskInfo = try await createRequestUseCase.execute(request: request)
                let response = TaskInfoMapper.toFlutterMap(taskInfo)
                result(response)
            } catch {
                result(FlutterError(code: "CREATE_REQUEST_FAILED", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func handleGetRequestStatus(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let requestId = args["requestId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Request ID is required", details: nil))
            return
        }
        
        Task {
            do {
                let taskInfo = try await getRequestStatusUseCase.execute(requestId: requestId)
                if let taskInfo = taskInfo {
                    let response = TaskInfoMapper.toFlutterMap(taskInfo)
                    result(response)
                } else {
                    result(nil)
                }
            } catch {
                result(FlutterError(code: "GET_STATUS_FAILED", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func handleGetBatchRequestStatus(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let requestIds = args["requestIds"] as? [String] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Request IDs list is required", details: nil))
            return
        }
        
        Task {
            do {
                var batchResult: [String: [String: Any]?] = [:]
                for requestId in requestIds {
                    let taskInfo = try await getRequestStatusUseCase.execute(requestId: requestId)
                    if let taskInfo = taskInfo {
                        batchResult[requestId] = TaskInfoMapper.toFlutterMap(taskInfo)
                    } else {
                        batchResult[requestId] = nil
                    }
                }
                result(batchResult)
            } catch {
                result(FlutterError(code: "GET_BATCH_STATUS_FAILED", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func handleGetResponse(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let requestId = args["requestId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Request ID is required", details: nil))
            return
        }
        
        Task {
            do {
                let taskInfo = try await getResponseUseCase.execute(requestId: requestId)
                if let taskInfo = taskInfo {
                    let response = TaskInfoMapper.toFlutterMap(taskInfo)
                    result(response)
                } else {
                    result(nil)
                }
            } catch {
                result(FlutterError(code: "GET_RESPONSE_FAILED", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func handleCancelRequest(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let requestId = args["requestId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Request ID is required", details: nil))
            return
        }
        
        Task {
            do {
                let cancelled = try await cancelRequestUseCase.execute(requestId: requestId)
                result(cancelled)
            } catch {
                result(FlutterError(code: "CANCEL_REQUEST_FAILED", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func handleDeleteRequest(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let requestId = args["requestId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Request ID is required", details: nil))
            return
        }
        
        Task {
            do {
                let deleted = try await deleteRequestUseCase.execute(requestId: requestId)
                result(deleted)
            } catch {
                result(FlutterError(code: "DELETE_REQUEST_FAILED", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func handleGetPendingTasks(call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            do {
                let pendingTasks = try await repository.getPendingTasks()
                let response = pendingTasks.map { task in
                    [
                        "requestId": task.requestId,
                        "registrationDate": task.registrationDate
                    ]
                }
                result(response)
            } catch {
                result(FlutterError(code: "GET_PENDING_TASKS_FAILED", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func handleCancelAllTasks(call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            do {
                let count = try await repository.cancelAllTasks()
                result(count)
            } catch {
                result(FlutterError(code: "CANCEL_ALL_TASKS_FAILED", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    // MARK: - Queue Management Methods
    
    private func handleGetQueueStats(call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            let stats = await repository.getQueueStats()
            result(stats.toDict())
        }
    }
    
    private func handleSetMaxConcurrentTasks(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let count = args["count"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "count is required", details: nil))
            return
        }
        
        Task {
            await repository.setMaxConcurrentTasks(count)
            result(true)
        }
    }
    
    private func handleSetMaxQueueSize(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let size = args["size"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "size is required", details: nil))
            return
        }
        
        Task {
            await repository.setMaxQueueSize(size)
            result(true)
        }
    }
    
    private func handleSyncQueueState(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // On iOS, synchronization is not required because we use an actor
        result(true)
    }
    
    private func handleProcessQueue(call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            await repository.processQueue()
            result(true)
        }
    }
    
    /// Gets pending completed tasks from UserDefaults queue and delivers them.
    /// This is a fallback for when the EventChannel couldn't deliver events directly.
    private func handleGetPendingCompletedTasks(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let pendingTasks = TaskCompletedEventStreamHandler.getPendingCompletedTasks()
        result(pendingTasks)
    }
}

File: android/src/main/kotlin/com/tsitser/background_http_plugin/presentation/handler/MethodCallHandler.kt
package com.tsitser.background_http_plugin.presentation.handler

import android.content.Context
import com.tsitser.background_http_plugin.data.mapper.RequestMapper
import com.tsitser.background_http_plugin.data.mapper.TaskInfoMapper
import com.tsitser.background_http_plugin.data.repository.TaskRepositoryImpl
import com.tsitser.background_http_plugin.domain.usecase.*
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Handler for method calls from Flutter.
 */
class MethodCallHandler(private val context: Context) : MethodChannel.MethodCallHandler {

    private val repository = TaskRepositoryImpl(context)
    private val createRequestUseCase = CreateRequestUseCase(repository)
    private val getRequestStatusUseCase = GetRequestStatusUseCase(repository)
    private val getResponseUseCase = GetResponseUseCase(repository)
    private val cancelRequestUseCase = CancelRequestUseCase(repository)
    private val deleteRequestUseCase = DeleteRequestUseCase(repository)

    private val scope = CoroutineScope(Dispatchers.Main)

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "createRequest" -> handleCreateRequest(call, result)
            "getRequestStatus" -> handleGetRequestStatus(call, result)
            "getBatchRequestStatus" -> handleGetBatchRequestStatus(call, result)
            "getResponse" -> handleGetResponse(call, result)
            "cancelRequest" -> handleCancelRequest(call, result)
            "deleteRequest" -> handleDeleteRequest(call, result)
            "getPendingTasks" -> handleGetPendingTasks(call, result)
            "cancelAllTasks" -> handleCancelAllTasks(call, result)
            // New methods for queue management
            "getQueueStats" -> handleGetQueueStats(call, result)
            "setMaxConcurrentTasks" -> handleSetMaxConcurrentTasks(call, result)
            "setMaxQueueSize" -> handleSetMaxQueueSize(call, result)
            "syncQueueState" -> handleSyncQueueState(call, result)
            "processQueue" -> handleProcessQueue(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleCreateRequest(call: MethodCall, result: MethodChannel.Result) {
        try {
            @Suppress("UNCHECKED_CAST")
            val requestMap = call.arguments as? Map<*, *>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request data is required", null)
                    return
                }

            scope.launch {
                try {
                    val request = RequestMapper.fromFlutterMap(requestMap)
                    val taskInfo = withContext(Dispatchers.IO) {
                        createRequestUseCase(request)
                    }
                    val response = TaskInfoMapper.toFlutterMap(taskInfo)
                    result.success(response)
                } catch (e: Exception) {
                    android.util.Log.e("MethodCallHandler", "Error creating request", e)
                    result.error("CREATE_REQUEST_FAILED", "${e.message}\n${e.stackTraceToString()}", null)
                }
            }
        } catch (e: Exception) {
            android.util.Log.e("MethodCallHandler", "Error parsing request", e)
            result.error("INVALID_ARGUMENT", "${e.message}\n${e.stackTraceToString()}", null)
        }
    }

    private fun handleGetRequestStatus(call: MethodCall, result: MethodChannel.Result) {
        try {
            @Suppress("UNCHECKED_CAST")
            val args = call.arguments as? Map<*, *>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request ID is required", null)
                    return
                }

            val requestId = args["requestId"] as? String
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request ID is required", null)
                    return
                }

            scope.launch {
                try {
                    val taskInfo = withContext(Dispatchers.IO) {
                        getRequestStatusUseCase(requestId)
                    }
                    if (taskInfo != null) {
                        val response = TaskInfoMapper.toFlutterMap(taskInfo)
                        result.success(response)
                    } else {
                        result.success(null)
                    }
                } catch (e: Exception) {
                    result.error("GET_STATUS_FAILED", e.message, null)
                }
            }
        } catch (e: Exception) {
            result.error("INVALID_ARGUMENT", e.message, null)
        }
    }

    private fun handleGetBatchRequestStatus(call: MethodCall, result: MethodChannel.Result) {
        try {
            @Suppress("UNCHECKED_CAST")
            val args = call.arguments as? Map<*, *>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request IDs are required", null)
                    return
                }

            @Suppress("UNCHECKED_CAST")
            val requestIds = args["requestIds"] as? List<*>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request IDs list is required", null)
                    return
                }

            scope.launch {
                try {
                    val batchResult = withContext(Dispatchers.IO) {
                        val resultMap = mutableMapOf<String, Map<String, Any>?>()
                        for (requestIdObj in requestIds) {
                            val requestId = requestIdObj as? String ?: continue
                            val taskInfo = getRequestStatusUseCase(requestId)
                            if (taskInfo != null) {
                                resultMap[requestId] = TaskInfoMapper.toFlutterMap(taskInfo)
                            } else {
                                resultMap[requestId] = null
                            }
                        }
                        resultMap
                    }
                    result.success(batchResult)
                } catch (e: Exception) {
                    result.error("GET_BATCH_STATUS_FAILED", e.message, null)
                }
            }
        } catch (e: Exception) {
            result.error("INVALID_ARGUMENT", e.message, null)
        }
    }

    private fun handleGetResponse(call: MethodCall, result: MethodChannel.Result) {
        try {
            @Suppress("UNCHECKED_CAST")
            val args = call.arguments as? Map<*, *>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request ID is required", null)
                    return
                }

            val requestId = args["requestId"] as? String
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request ID is required", null)
                    return
                }

            scope.launch {
                try {
                    val taskInfo = withContext(Dispatchers.IO) {
                        getResponseUseCase(requestId)
                    }
                    if (taskInfo != null) {
                        val response = TaskInfoMapper.toFlutterMap(taskInfo)
                        result.success(response)
                    } else {
                        result.success(null)
                    }
                } catch (e: Exception) {
                    result.error("GET_RESPONSE_FAILED", e.message, null)
                }
            }
        } catch (e: Exception) {
            result.error("INVALID_ARGUMENT", e.message, null)
        }
    }

    private fun handleCancelRequest(call: MethodCall, result: MethodChannel.Result) {
        try {
            @Suppress("UNCHECKED_CAST")
            val args = call.arguments as? Map<*, *>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request ID is required", null)
                    return
                }

            val requestId = args["requestId"] as? String
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request ID is required", null)
                    return
                }

            scope.launch {
                try {
                    val cancelled = withContext(Dispatchers.IO) {
                        cancelRequestUseCase(requestId)
                    }
                    result.success(cancelled)
                } catch (e: Exception) {
                    result.error("CANCEL_REQUEST_FAILED", e.message, null)
                }
            }
        } catch (e: Exception) {
            result.error("INVALID_ARGUMENT", e.message, null)
        }
    }

    private fun handleDeleteRequest(call: MethodCall, result: MethodChannel.Result) {
        try {
            @Suppress("UNCHECKED_CAST")
            val args = call.arguments as? Map<*, *>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request ID is required", null)
                    return
                }

            val requestId = args["requestId"] as? String
                ?: run {
                    result.error("INVALID_ARGUMENT", "Request ID is required", null)
                    return
                }

            scope.launch {
                try {
                    val deleted = withContext(Dispatchers.IO) {
                        deleteRequestUseCase(requestId)
                    }
                    result.success(deleted)
                } catch (e: Exception) {
                    result.error("DELETE_REQUEST_FAILED", e.message, null)
                }
            }
        } catch (e: Exception) {
            result.error("INVALID_ARGUMENT", e.message, null)
        }
    }

    private fun handleGetPendingTasks(call: MethodCall, result: MethodChannel.Result) {
        scope.launch {
            try {
                val pendingTasks = withContext(Dispatchers.IO) {
                    repository.getPendingTasks()
                }
                val response = pendingTasks.map { task ->
                    mapOf(
                        "requestId" to task.requestId,
                        "registrationDate" to task.registrationDate
                    )
                }
                result.success(response)
            } catch (e: Exception) {
                android.util.Log.e("MethodCallHandler", "Error getting pending tasks", e)
                result.error("GET_PENDING_TASKS_FAILED", e.message, null)
            }
        }
    }

    private fun handleCancelAllTasks(call: MethodCall, result: MethodChannel.Result) {
        scope.launch {
            try {
                val count = withContext(Dispatchers.IO) {
                    repository.cancelAllTasks()
                }
                result.success(count)
            } catch (e: Exception) {
                android.util.Log.e("MethodCallHandler", "Error cancelling all tasks", e)
                result.error("CANCEL_ALL_TASKS_FAILED", e.message, null)
            }
        }
    }
    
    private fun handleGetQueueStats(call: MethodCall, result: MethodChannel.Result) {
        try {
            val stats = repository.getQueueStats()
            result.success(stats.toMap())
        } catch (e: Exception) {
            android.util.Log.e("MethodCallHandler", "Error getting queue stats", e)
            result.error("GET_QUEUE_STATS_FAILED", e.message, null)
        }
    }
    
    private fun handleSetMaxConcurrentTasks(call: MethodCall, result: MethodChannel.Result) {
        try {
            @Suppress("UNCHECKED_CAST")
            val args = call.arguments as? Map<*, *>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Arguments are required", null)
                    return
                }
            
            val count = (args["count"] as? Number)?.toInt()
                ?: run {
                    result.error("INVALID_ARGUMENT", "count is required", null)
                    return
                }
            
            scope.launch {
                try {
                    withContext(Dispatchers.IO) {
                        repository.setMaxConcurrentTasks(count)
                    }
                    result.success(true)
                } catch (e: Exception) {
                    result.error("SET_MAX_CONCURRENT_FAILED", e.message, null)
                }
            }
        } catch (e: Exception) {
            result.error("INVALID_ARGUMENT", e.message, null)
        }
    }
    
    private fun handleSetMaxQueueSize(call: MethodCall, result: MethodChannel.Result) {
        try {
            @Suppress("UNCHECKED_CAST")
            val args = call.arguments as? Map<*, *>
                ?: run {
                    result.error("INVALID_ARGUMENT", "Arguments are required", null)
                    return
                }
            
            val size = (args["size"] as? Number)?.toInt()
                ?: run {
                    result.error("INVALID_ARGUMENT", "size is required", null)
                    return
                }
            
            repository.setMaxQueueSize(size)
            result.success(true)
        } catch (e: Exception) {
            result.error("SET_MAX_QUEUE_SIZE_FAILED", e.message, null)
        }
    }
    
    private fun handleSyncQueueState(call: MethodCall, result: MethodChannel.Result) {
        scope.launch {
            try {
                withContext(Dispatchers.IO) {
                    repository.syncQueueState()
                }
                result.success(true)
            } catch (e: Exception) {
                android.util.Log.e("MethodCallHandler", "Error syncing queue state", e)
                result.error("SYNC_QUEUE_STATE_FAILED", e.message, null)
            }
        }
    }
    
    private fun handleProcessQueue(call: MethodCall, result: MethodChannel.Result) {
        scope.launch {
            try {
                withContext(Dispatchers.IO) {
                    repository.processQueue()
                }
                result.success(true)
            } catch (e: Exception) {
                android.util.Log.e("MethodCallHandler", "Error processing queue", e)
                result.error("PROCESS_QUEUE_FAILED", e.message, null)
            }
        }
    }
}

